libsolidity/inlineasm/AsmParser.cpp

```diff
@@ -606,7 +606,7 @@ bool Parser::isValidNumberLiteral(string const& _literal)
 {
 	try
 	{
-		u256(_literal);
+		u256 _literal;
```

Comment:
I think this is a bad idea, since this just creates a variable called `_literal` shadowing the other one. The goal here is to try converting it to `u256`.

Reply:
Yeah you're right.  I was a little confused by the make output:
```
/mnt/data/repos/solidity/libsolidity/inlineasm/AsmParser.cpp: In static member function ‘static bool dev::solidity::assembly::Parser::isValidNumberLiteral(const string&)’:
/mnt/data/repos/solidity/libsolidity/inlineasm/AsmParser.cpp:609:7: error: unnecessary parentheses in declaration of ‘_literal’ [-Werror=parentheses]
   u256(_literal);
       ^
cc1plus: all warnings being treated as errors
make[2]: *** [libsolidity/CMakeFiles/solidity.dir/build.make:583: libsolidity/CMakeFiles/solidity.dir/inlineasm/AsmParser.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:274: libsolidity/CMakeFiles/solidity.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

Strangely replacing it with 
```
auto k = u256(_literal);
(void) k;
```
works without throwing an error, but there most likely is a better solution :P

Reply:
Also since this change didn't trigger any test failure that suggests we need to add a test for this :)

Reply:
I'm not familiar with the boost multiprecision lib, but couldn't it be that the code above does not do what you intended it to do?


I checked (disabling Werror) the original code with a sample string

```c++
	std::string asdf = "i am not a number at all";
	try
	{
		u256(asdf);
	}
	catch (...)
	{
		std:: cerr << "works.\n";
		return false;
	}
	std :: cerr << "does not work.\n";
	std::exit(-1);
```
output
```
does not work.
```
and it prints
"does not work" to the cli, thus recognizes the string as a number. Or did I missunderstand which strings can be converted?

Reply:
and the version  I proposed (replacing it with auto k = uint256(asdf); (void) k works; thus throws an exception.

Reply:
If an exception is caught (and as a result `false` is returned) that means it was a non-number input. I think the code is proper.

Not sure why it accepts the string without issues in your case. A test case what would be accepted by the parser, but not by this function, `1e200` or `1.200`.

Reply:
But isn't it strange that the code 
```
u256(asdf);
```c++
had the exact same output as the wrong code that I committed to this pull request
```c++
u256 asdf; 
```
whereas
```c++
auto _ = u256(asdf);
(void)_;
```
reacts differently and throws an exception to the string I tested it with?

Reply:
I double checked the code:
- there is a test case already
- the reason it still works even if you disable it, because there's a manual check `_literal.find_first_not_of` outside the try/catch statement

I think the best would be 
```
auto tmp = u256(_literal);
(void)tmp;
```